proposed/2021/CentralPackageManagementTransitive.md

```diff
@@ -0,0 +1,201 @@
+```
+Note that since **`PackageC` is no longer considered transitive, it must now be elevated to be a top-level dependency of my package**.  This is because `ClassLibrary1` was built against version `5.0.0` of `PackageC` so the transitive dependencies of `PackageA` and `PackageB` are no longer accurate.  The act of "pinning" a version of a transitive dependency effectively transformed it into an explicit dependency.
+
+## Referencing packages with CPVM
+
+To manage the package versions in one location, the `Version` metadata is removed from `PackageReference` items in the project:
+
+**`ClassLibrary1.csproj`**
+```diff
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+  </PropertyGroup>
+  <ItemGroup>
+-   <PackageReference Include="PackageA" Version="1.0.0" />
++   <PackageReference Include="PackageA" />
+-   <PackageReference Include="PackageB" Version="2.0.0" />
++   <PackageReference Include="PackageB" />
+  </ItemGroup>
+</Project>
+```
+
+And a `Directory.Packages.props` file is created with the following content:
+
+**`Directory.Packages.props`**
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="1.0.0" />
+    <PackageVersion Include="PackageB" Version="2.0.0" />
+  </ItemGroup>
+</Project>
+```
+
+All projects in the repository that reference `PackageA` or `PackageB` explicitly will use the specified value in a corresponding `PackageVersion` item.
+
+If a user wants to override the transitive version of `PackageC`, they must add an explicit `PackageReference` item:
+```diff
+<ItemGroup>
+  <PackageReference Include="PackageA" />
+  <PackageReference Include="PackageB" />
++ <PackageReference Include="PackageC" />
+</ItemGroup>
+```
+
+And add a corresponding `PackageVersion` item to Directory.Packages.props:
+```diff
+<ItemGroup>
+  <PackageVersion Include="PackageA" Version="1.0.0" />
+  <PackageVersion Include="PackageB" Version="2.0.0" />
++ <PackageVersion Include="PackageC" Version="5.0.0" />
+</ItemGroup>
+```
+
+The result is the same as before where the resolved `PackageC` is version **`5.0.0`** and affects the restored packages, reference assemblies passed to the compiler, files copied to the output directory, and dependencies in used during package creation.  The transitive dependency is no longer considered transitive and instead is an explicit dependency.
+
+### Proposal to allow transitive version selection based on specified PackageVersion items
+In the previous example, a user must specify an explicit `PackageReference` to a package to override the transitive version.  This elevates the reference from transitive to explicit and behaves as a user expects.  However, some users want a `PackageVersion` to automatically override a transitive version even without an explicit top-level `PackageReference`.  The reason for this feature request is for better **scale** and **performance**.
```

Comment:
I disagree that the only reason for this request is scale and performance. 

We have multiple top level projects in a solution, and due to some deployment complexity, install them side by side in the same folder. We want the dependencies to all be on the same version, even when Nuget is not able to resolve across dependency trees. Otherwise, we might end up with a dll of unknown version depending on which build finished first.

Reply:
Going by the subsequent paragraph, I think @JonDouglas is using the term "scale" in the sense of "managing the versions in every .csproj individually doesn't _scale_", That seems to largely align with your use-case, in that you _can_ do what you need now, but it's an unreasonable amount of work. 

I'd certainly agree that all of that meaning isn't immediately clear from the phrase "better **scale** and **performance**" (although I can't immediately think of better wording). It made me think of the scale of the dependency graph, or something like that.

Reply:
Thank you both. Yes the wording can use some word smithing. It was a hidden benefit that users use this feature today for. We just want to express that perf isn't any guarantee here and especially as projects scale(become larger), it is even less of a guarantee. Rather the focus is on UX, convenience, and more.

Reply:
I'd definitely add `unification` as the primary purpose or something along the lines that @jjmanton is suggesting.

Reply:
Thanks again everybody for tackling this issue! It's really not trivial to get this right.

To give a (hopefully) easy-to-understand example of what `unification` could mean:

- I have several plug-ins projects meant for a common main App.
- They are built separately, but need to run together in the same app domain.
- Plug-Ins will depend on App, but they are not allowed to depend on each other.
- Each plug-in might use any external nuget package reference it likes.

Transitive pinning would give me a central file where to synchronize the (transitive) dependencies of all plug-ins to common versions. This can then be collected into the Main App's `assemblyBinding` version override list, and I am done.

To find conflicting version issues before they hit at runtime, I can do a simple run of `dotnet list package <PROJ> --include-transitive` over all plug-ins and the main App, and compare the resolved versions. If there is a conflict, I can resolve it centrally to a common version.

### Another advantage of TP against elevation: Detect and remove no-longer-needed dependencies 
Another (I think) important point where elevation of transitive packages becomes unmanageable, is when you think about how your App evolves over time:

If I have to elevate conflicting packages to top-level dependencies, I wouldn't know how to detect when it becomes unneeded in a later version of that 3rd-party-package. Thus, I would *never be able to get rid of it again*. With transitive pinning, it would disappear automatically from my dependency graph and the compile folder, and I can check for that and remove the pinned version specification at my convenience.

### Minor point: Managing the stability of your app's full DLL set
When transitive pinning is used together with lock-files and a transitive company nuget mirror, I can conveniently manage and _guarantee_ DLL stability between re-builds (e.g. on CI infrastructure), which I think is vital against accidental or malicious changes (= new nuget version releases) in the supply-chain-graph.